### PR TITLE
[search] Use MemTrie for street synonyms.

### PR DESCRIPTION
**Pull request by ygorshenin**
@mpimenov @bykoianko PTAL

### CODE REVIEW

File: indexer/search_string_utils.cpp [hunk 1]
```diff
@@ -94,11 +96,67 @@ char const * kStreetTokensSeparator = "\t -,.";
...
+
+      created = true;
+      m_subtrees.emplace_back(c, make_unique<SubTree>());
+      return *m_subtrees.back().second;
+    }
+
+    void Clear() { m_subtrees.clear(); }
+
+  private:
+    buffer_vector<pair<Char, std::unique_ptr<SubTree>>, 8> m_subtrees;
```
Comment (bykoianko) [inline indexer/search_string_utils.cpp]: Do you really need buffer_vector in this case? vector?

Comment (ygorshenin) [inline indexer/search_string_utils.cpp]: Yes, I need. Please, take a look at the old implementation of this class - it used buffer_vector in nodes, as the branching factor is quite low, therefore it's ok to pay for unused space.

File: base/mem_trie.hpp [hunk 2]
```diff
@@ -157,10 +212,9 @@ class MemTrie
     auto const * cur = &m_root;
     for (auto const & c : key)
     {
-      auto const it = cur->m_moves.find(c);
-      if (it == cur->m_moves.end())
-        return nullptr;
-      cur = it->second.get();
+      cur = cur->m_moves.GetSubTree(c);
```
Comment (mpimenov) [inline base/mem_trie.hpp]: s/SubTree/Subtree/g

File: indexer/search_string_utils.cpp [hunk 3]
```diff
@@ -94,11 +96,67 @@ char const * kStreetTokensSeparator = "\t -,.";
...
-  using TStrings = search::StringSet<UniChar, 8>;
+  struct BooleanSum
+  {
+    void Add(bool value) { m_value = m_value || value; }
+
+    template <typename ToDo>
+    void ForEach(ToDo && toDo) const
+    {
+      if (m_value)
+        toDo(m_value);
```
Comment (mpimenov) [inline indexer/search_string_utils.cpp]: So this is toDo(true)?
I think that either the check should be moved to the client (as in ```toDo(m_value)```) or the argument should be removed.

File: indexer/search_string_utils.cpp [hunk 4]
```diff
@@ -94,11 +96,67 @@ char const * kStreetTokensSeparator = "\t -,.";
...
+    }
+
+    void Clear() { m_value = false; }
+
+    bool m_value = false;
+  };
+
+  template <typename Char, typename SubTree>
+  class Moves
```
Comment (mpimenov) [inline indexer/search_string_utils.cpp]: Probably renaming this to BufferVectorMoves and moving closer to MapMoves is cleaner.

Comment (ygorshenin) [inline indexer/search_string_utils.cpp]: This is quite custom implementation of branching, I really don't want to expose it out of this translation unit.

File: indexer/search_string_utils.cpp (hunk 3 above)
Comment (ygorshenin) [inline indexer/search_string_utils.cpp]: Unfortunately I don't see an easy way to get rid of the argument.

Comment (ygorshenin) [inline indexer/search_string_utils.cpp]: Moreover, without value you won't be able to distinguish between internal nodes and nodes that mean end of some string.

File: base/mem_trie.hpp (hunk 2 above)
Comment (ygorshenin) [inline base/mem_trie.hpp]: Done

**Comment by ygorshenin** [comment]
@mpimenov @bykoianko PTAL

File: indexer/search_string_utils.cpp [hunk 5]
```diff
@@ -168,24 +225,32 @@ class StreetsSynonymsHolder
     for (auto const * s : affics)
     {
       UniString const us = NormalizeAndSimplifyString(s);
-      m_strings.Add(us.begin(), us.end());
+      m_strings.Add(us, true);
```
Comment (kshalnev) [inline indexer/search_string_utils.cpp]: add comment /* who is true */?

File: base/mem_trie.hpp [hunk 6]
```diff
@@ -5,15 +5,76 @@
...
+  std::map<Char, std::unique_ptr<Subtree>> m_subtrees;
+};
+
+template <typename Value>
+struct VectorValues
+{
+  template <typename V>
+  void Add(V && v)
+  {
+    m_values.push_back(std::forward<V>(v));
```
Comment (kshalnev) [inline base/mem_trie.hpp]: emplace_back?

File: indexer/search_string_utils.cpp (hunk 3 above)
Comment (mpimenov) [inline indexer/search_string_utils.cpp]: How does the value help you? It is always true as far as toDo is concerned.

File: base/mem_trie.hpp [hunk 7]
```diff
@@ -170,8 +224,7 @@ class MemTrie
   template <typename ToDo>
   void ForEachInNode(Node const & node, String const & prefix, ToDo && toDo) const
   {
-    for (auto const & value : node.m_values)
-      toDo(prefix, value);
+    node.m_values.ForEach(std::bind(std::forward<ToDo>(toDo), prefix, std::placeholders::_1));
```
Comment (kshalnev) [inline base/mem_trie.hpp]: ref(prefix)? чтоб не копировать?

File: indexer/search_string_utils.cpp (hunk 3 above)
Comment (ygorshenin) [inline indexer/search_string_utils.cpp]: Please, take a look at the current version of this function. If is removed.

File: indexer/search_string_utils.cpp (hunk 5 above)
Comment (ygorshenin) [inline indexer/search_string_utils.cpp]: Done

File: base/mem_trie.hpp (hunk 6 above)
Comment (ygorshenin) [inline base/mem_trie.hpp]: Done

File: base/mem_trie.hpp (hunk 7 above)
Comment (ygorshenin) [inline base/mem_trie.hpp]: Errr, unfortunately we don't use cpp14, that's why I can't use auto in lambdas. But in the case of bind, thanks, it's a right thing to use ref here.

**Comment by ygorshenin** [comment]
@bykoianko @kshalnev PTAL

**Comment by kshalnev** [comment]
LGTM

**Pull request merged**